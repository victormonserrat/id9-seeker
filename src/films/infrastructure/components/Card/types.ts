import { HTMLAttributes } from 'react'

export type FilmCardProps = Omit<HTMLAttributes<HTMLDivElement>, 'title'> & {
  body?: string
  coverUrl: string
  episode: string
  release: string
  title: string
}

export type UseFilmCardProps = {
  release: FilmCardProps['release']
}
