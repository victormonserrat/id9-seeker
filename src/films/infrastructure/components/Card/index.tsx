import Image from 'next/image'
import { memo } from 'react'

import {
  Body,
  Component,
  Content,
  ImageContainer,
  Release,
  Separator,
  Title,
} from './styles'
import { FilmCardProps } from './types'

const FilmCard = ({
  body,
  coverUrl,
  episode,
  release,
  title,
  ...props
}: FilmCardProps) => (
  <Component {...props}>
    <ImageContainer>
      <Image
        alt={`${title} Cover`}
        height="50%"
        layout="responsive"
        priority
        src={coverUrl}
        width="100%"
      />
    </ImageContainer>
    <Separator aria-hidden role="separator" />
    <Content>
      <div>
        <Title>{`Episode ${episode} - ${title}`}</Title>
        {body ? <Body>{body}</Body> : null}
      </div>
      <Release>{release}</Release>
    </Content>
  </Component>
)

export default memo(FilmCard)

export type { FilmCardProps }
