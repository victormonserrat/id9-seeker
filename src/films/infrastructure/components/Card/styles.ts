import styled from 'styled-components'

import {
  bottomEdgeStyles,
  leftEdgeStyles,
  revertBottomEdgeStyles,
} from '~/common/infrastructure/styles/edges'
import media from '~/common/infrastructure/styles/media'

export const Body = styled.p`
  display: none;

  ${media.large} {
    color: ${({ theme }) => theme.colors.films.card.body};
    display: revert;
    font-size: ${({ theme }) => theme.font.sizes.small};
    font-weight: ${({ theme }) => theme.font.weights.normal};
    line-height: ${({ theme }) => theme.font.sizes.medium};
    margin-top: 0.375rem;
  }
`

export const Title = styled.h5`
  color: ${({ theme }) => theme.colors.films.card.title.default};
  font-size: ${({ theme }) => theme.font.sizes.medium};
  font-weight: ${({ theme }) => theme.font.weights.medium};
  text-transform: uppercase;

  ${media.medium} {
    font-size: ${({ theme }) => theme.font.sizes.large};
  }
`

export const Component = styled.div`
  background-color: ${({ theme }) => theme.colors.films.card.background};
  border-radius: 0.375rem;
  display: flex;
  flex-direction: column;
  ${({ theme }) => bottomEdgeStyles(theme.colors.films.card.background)}

  ${media.medium} {
    ${revertBottomEdgeStyles}
    ${({ theme }) => leftEdgeStyles(theme.colors.films.card.background)}
    flex-direction: row-reverse;
    width: 100%;
  }
`

export const Content = styled.div`
  background-image: url(https://static-mh.content.disney.io/starwars/assets/shared/bg_hash_top-dca2c5ab1b2e.png);
  background-size: 0.4375rem;
  background-repeat: repeat-x;
  background-position: top left;
  display: flex;
  flex-direction: column;
  gap: 2rem;
  justify-content: space-between;
  padding: 0.75rem;

  ${media.medium} {
    width: calc(55% - 0.05rem);
  }
`

export const ImageContainer = styled.div`
  img {
    border-radius: 0.375rem 0.375rem 0 0;
  }

  ${media.medium} {
    width: calc(45% - 0.05rem);

    img {
      border-radius: 0 0.375rem 0.375rem 0;
    }
  }
`

export const Release = styled.span`
  color: ${({ theme }) => theme.colors.films.card.release};
  font-size: ${({ theme }) => theme.font.sizes.small};
  font-weight: ${({ theme }) => theme.font.weights.normal};
  text-transform: uppercase;
`

export const Separator = styled.div`
  background-color: ${({ theme }) => theme.colors.films.card.separator};
  height: 0.1rem;

  ${media.medium} {
    height: revert;
    width: 0.1rem;
  }
`
