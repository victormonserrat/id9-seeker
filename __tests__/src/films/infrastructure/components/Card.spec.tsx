import Card from '~/films/infrastructure/components/Card'
import { render } from '~/tests/utils'

describe('Film Card', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Card
        coverUrl="https://cover.url"
        episode="Episode"
        title="Title"
        release="Release"
      />,
    )

    expect(container).toMatchSnapshot('Regular')
  })

  it('renders with body correctly', () => {
    const { container } = render(
      <Card
        coverUrl="https://cover.url"
        episode="Episode"
        title="Title"
        release="Release"
        body="Body"
      />,
    )

    expect(container).toMatchSnapshot('With body')
  })
})
