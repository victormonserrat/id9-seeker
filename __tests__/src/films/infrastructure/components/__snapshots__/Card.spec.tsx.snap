// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Film Card renders correctly: Regular 1`] = `
<div>
  <div
    class="sc-af21c34c-2 gGVSDn"
  >
    <div
      class="sc-af21c34c-4 gkgsfe"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 50% 0px 0px 0px;"
        />
        <img
          alt="Title Cover"
          data-nimg="responsive"
          decoding="async"
          sizes="100vw"
          src="/_next/image?url=https%3A%2F%2Fcover.url&w=3840&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fcover.url&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fcover.url&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fcover.url&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fcover.url&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fcover.url&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fcover.url&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fcover.url&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fcover.url&w=3840&q=75 3840w"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
      </span>
    </div>
    <div
      aria-hidden="true"
      class="sc-af21c34c-6 dHhRXG"
      role="separator"
    />
    <div
      class="sc-af21c34c-3 ipzcKO"
    >
      <div>
        <h5
          class="sc-af21c34c-1 hgDvBZ"
        >
          Episode Episode - Title
        </h5>
      </div>
      <span
        class="sc-af21c34c-5 gkBrxH"
      >
        Release
      </span>
    </div>
  </div>
</div>
`;

exports[`Film Card renders with body correctly: With body 1`] = `
<div>
  <div
    class="sc-af21c34c-2 gGVSDn"
  >
    <div
      class="sc-af21c34c-4 gkgsfe"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 50% 0px 0px 0px;"
        />
        <img
          alt="Title Cover"
          data-nimg="responsive"
          decoding="async"
          sizes="100vw"
          src="/_next/image?url=https%3A%2F%2Fcover.url&w=3840&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fcover.url&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fcover.url&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fcover.url&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fcover.url&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fcover.url&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fcover.url&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fcover.url&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fcover.url&w=3840&q=75 3840w"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
      </span>
    </div>
    <div
      aria-hidden="true"
      class="sc-af21c34c-6 dHhRXG"
      role="separator"
    />
    <div
      class="sc-af21c34c-3 ipzcKO"
    >
      <div>
        <h5
          class="sc-af21c34c-1 hgDvBZ"
        >
          Episode Episode - Title
        </h5>
        <p
          class="sc-af21c34c-0 AFxMt"
        >
          Body
        </p>
      </div>
      <span
        class="sc-af21c34c-5 gkBrxH"
      >
        Release
      </span>
    </div>
  </div>
</div>
`;
